File: apps/api/src/common/crypto.ts
import { createCipheriv, createDecipheriv, randomBytes } from 'crypto';

const ALGO = 'aes-256-gcm';

function key(): Buffer {
  const hex = process.env.CONFIG_ENCRYPTION_KEY;
  if (!hex || hex.length !== 64) {
    throw new Error('CONFIG_ENCRYPTION_KEY must be a 32-byte hex string (openssl rand -hex 32)');
  }
  return Buffer.from(hex, 'hex');
}

export function encryptJson(value: unknown): string {
  const iv = randomBytes(12);
  const cipher = createCipheriv(ALGO, key(), iv);
  const plaintext = Buffer.from(JSON.stringify(value), 'utf8');
  const enc = Buffer.concat([cipher.update(plaintext), cipher.final()]);
  const tag = cipher.getAuthTag();
  return Buffer.concat([iv, tag, enc]).toString('base64');
}

export function decryptJson<T = unknown>(payload: string): T {
  const buf = Buffer.from(payload, 'base64');
  const iv = buf.subarray(0, 12);
  const tag = buf.subarray(12, 28);
  const enc = buf.subarray(28);
  const decipher = createDecipheriv(ALGO, key(), iv);
  decipher.setAuthTag(tag);
  const dec = Buffer.concat([decipher.update(enc), decipher.final()]);
  return JSON.parse(dec.toString('utf8')) as T;
}

export function hasEncryptionKey(): boolean {
  const hex = process.env.CONFIG_ENCRYPTION_KEY;
  return !!hex && hex.length === 64;
}

File: apps/api/internal/crypto/crypto.go
package crypto

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"io"
)

const (
	ivLen  = 12
	tagLen = 16
)

func keyFromHex(hexKey string) ([]byte, error) {
	if len(hexKey) != 64 {
		return nil, errors.New("CONFIG_ENCRYPTION_KEY must be a 32-byte hex string (openssl rand -hex 32)")
	}
	key, err := hex.DecodeString(hexKey)
	if err != nil {
		return nil, fmt.Errorf("CONFIG_ENCRYPTION_KEY is not valid hex: %w", err)
	}
	return key, nil
}

func HasEncryptionKey(hexKey string) bool {
	return len(hexKey) == 64
}

func EncryptJSON(hexKey string, value any) (string, error) {
	key, err := keyFromHex(hexKey)
	if err != nil {
		return "", err
	}
	block, err := aes.NewCipher(key)
	if err != nil {
		return "", err
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return "", err
	}
	iv := make([]byte, ivLen)
	if _, err := io.ReadFull(rand.Reader, iv); err != nil {
		return "", err
	}
	plaintext, err := json.Marshal(value)
	if err != nil {
		return "", err
	}
	sealed := gcm.Seal(nil, iv, plaintext, nil)
	ct := sealed[:len(sealed)-tagLen]
	tag := sealed[len(sealed)-tagLen:]

	out := make([]byte, 0, ivLen+tagLen+len(ct))
	out = append(out, iv...)
	out = append(out, tag...)
	out = append(out, ct...)
	return base64.StdEncoding.EncodeToString(out), nil
}

func DecryptJSON(hexKey string, payload string, out any) error {
	key, err := keyFromHex(hexKey)
	if err != nil {
		return err
	}
	buf, err := base64.StdEncoding.DecodeString(payload)
	if err != nil {
		return fmt.Errorf("crypto: invalid base64 payload: %w", err)
	}
	if len(buf) < ivLen+tagLen {
		return errors.New("crypto: payload too short")
	}
	iv := buf[:ivLen]
	tag := buf[ivLen : ivLen+tagLen]
	ct := buf[ivLen+tagLen:]

	block, err := aes.NewCipher(key)
	if err != nil {
		return err
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return err
	}
	sealed := make([]byte, 0, len(ct)+tagLen)
	sealed = append(sealed, ct...)
	sealed = append(sealed, tag...)

	plaintext, err := gcm.Open(nil, iv, sealed, nil)
	if err != nil {
		return fmt.Errorf("crypto: decryption failed: %w", err)
	}
	return json.Unmarshal(plaintext, out)
}
